plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/LocationCompassEngine.kt

```diff
@@ -0,0 +1,300 @@
+            worldAxisForDeviceAxisX = SensorManager.AXIS_MINUS_Y
+            worldAxisForDeviceAxisY = SensorManager.AXIS_MINUS_X
+          }
+          Surface.ROTATION_180 -> {
+            worldAxisForDeviceAxisX = SensorManager.AXIS_MINUS_X
+            worldAxisForDeviceAxisY = SensorManager.AXIS_Y
+          }
+          Surface.ROTATION_270 -> {
+            worldAxisForDeviceAxisX = SensorManager.AXIS_Y
+            worldAxisForDeviceAxisY = SensorManager.AXIS_X
+          }
+          Surface.ROTATION_0 -> {
+            worldAxisForDeviceAxisX = SensorManager.AXIS_X
+            worldAxisForDeviceAxisY = SensorManager.AXIS_MINUS_Y
+          }
+          else -> {
+            worldAxisForDeviceAxisX = SensorManager.AXIS_X
+            worldAxisForDeviceAxisY = SensorManager.AXIS_MINUS_Y
+          }
+        }
+      }
+    }
+    SensorManager.remapCoordinateSystem(
+      rotationMatrix, worldAxisForDeviceAxisX,
+      worldAxisForDeviceAxisY, adjustedRotationMatrix
+    )
+
+    // Transform rotation matrix into azimuth/pitch/roll
+    SensorManager.getOrientation(adjustedRotationMatrix, orientation)
+
+    // The x-axis is all we care about here.
+    notifyCompassChangeListeners(Math.toDegrees(orientation[0].toDouble()).toFloat())
+
+    // Update the compassUpdateNextTimestamp
+    compassUpdateNextTimestamp = currentTime + LocationComponentConstants.COMPASS_UPDATE_RATE_MS
+  }
+
+  private fun notifyCompassChangeListeners(heading: Float) {
+    for (compassListener in compassListeners) {
+      compassListener.onCompassChanged(heading)
+    }
+    lastHeading = heading
+  }
+
+  private fun registerSensorListeners() {
+    if (isCompassSensorAvailable()) {
+      // Does nothing if the sensors already registered.
+      sensorManager.registerListener(
+        this,
+        compassSensor,
+        SENSOR_DELAY_MICROS
+      )
+    } else {
+      sensorManager.registerListener(
+        this,
+        gravitySensor,
+        SENSOR_DELAY_MICROS
+      )
+      sensorManager.registerListener(
+        this,
+        magneticFieldSensor,
+        SENSOR_DELAY_MICROS
+      )
+    }
+  }
+
+  private fun unregisterSensorListeners() {
+    if (isCompassSensorAvailable()) {
+      sensorManager.unregisterListener(this, compassSensor)
+    } else {
+      sensorManager.unregisterListener(this, gravitySensor)
+      sensorManager.unregisterListener(this, magneticFieldSensor)
+    }
+  }
+
+  private fun isCompassSensorAvailable() = compassSensor != null
+
+  /**
+   * Helper function, that filters newValues, considering previous values
+   *
+   * @param newValues      array of float, that contains new data
+   * @param smoothedValues array of float, that contains previous state
+   * @return float filtered array of float
+   */
+  private fun lowPassFilter(newValues: FloatArray, smoothedValues: FloatArray?): FloatArray {
+    if (smoothedValues == null) {
+      return newValues
+    }
+    for (i in newValues.indices) {
```

Comment:
is it possible that `newValues.size` != `smoothedValues.size`?

Reply:
No, from the doc the value size for `Sensor.TYPE_ACCELEROMETER`  and `Sensor.TYPE_MAGNETIC_FIELD` are 3.